charger/dadapower.go

```diff
@@ -0,0 +1,242 @@
+	dadapowerRegIdentification      = 1040 // 20
+)
+
+var dadapowerRegCurrents = []uint16{1006, 1007, 1008}
+
+// Dadapower charger implementation
+type Dadapower struct {
+	log       *util.Logger
+	conn      *modbus.Connection
+	regOffset uint16
+	current   uint16
+}
+
+func init() {
+	registry.Add("dadapower", NewDadapowerFromConfig)
+}
+
+// NewDadapowerFromConfig creates a Dadapower charger from generic config
+func NewDadapowerFromConfig(other map[string]interface{}) (api.Charger, error) {
+	cc := modbus.TcpSettings{}
+
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+
+	return NewDadapower(cc.URI, cc.ID)
+}
+
+// NewDadapower creates a Dadapower charger
+func NewDadapower(uri string, id uint8) (*Dadapower, error) {
+	conn, err := modbus.NewConnection(uri, "", "", 0, modbus.Tcp, id)
+	if err != nil {
+		return nil, err
+	}
+
+	log := util.NewLogger("dadapower")
+	conn.Logger(log.TRACE)
+
+	wb := &Dadapower{
+		log:     log,
+		conn:    conn,
+		current: 6,
+	}
+
+	// 5min failsafe timeout
+	if _, err := wb.conn.WriteSingleRegister(dadapowerRegFailsafeTimeout, 5*60); err != nil {
+		return nil, fmt.Errorf("could not set failsafe timeout: %v", err)
+	}
+
+	// The charging station may have multiple charging ports - use offset for register addresses for each port
+	if id > 1 {
+		wb.regOffset = (uint16(id) - 1) * 1000
+	}
+
+	go wb.heartbeat()
+
+	return wb, nil
+}
+
+// heartbeat implements the api.ChargerEx interface
+func (wb *Dadapower) heartbeat() {
+	for range time.NewTicker(time.Minute).C {
+		if _, err := wb.conn.ReadInputRegisters(dadapowerRegFailsafeTimeout, 1); err != nil {
+			wb.log.ERROR.Println("heartbeat:", err)
+		}
+	}
+}
+
+// Status implements the api.Charger interface
+func (wb *Dadapower) Status() (api.ChargeStatus, error) {
+	b, err := wb.conn.ReadInputRegisters(dadapowerRegPlugState+wb.regOffset, 1)
+
+	if err != nil {
+		return api.StatusNone, err
+	}
+
+	switch binary.BigEndian.Uint16(b) {
+	case 0x0A: // ready
+		return api.StatusA, nil
+	case 0x0B: // EV is present
+		return api.StatusB, nil
+	case 0x0C: // charging
+		return api.StatusC, nil
+	case 0x0D: // charging with ventilation
+		return api.StatusD, nil
+	case 0x0E: // failure (e.g. diode check, RCD failure)
+		return api.StatusE, nil
+	default:
+		return api.StatusNone, errors.New("invalid response")
+	}
+}
+
+// Enabled implements the api.Charger interface
+func (wb *Dadapower) Enabled() (bool, error) {
+	b, err := wb.conn.ReadInputRegisters(dadapowerRegChargingAllowed+wb.regOffset, 1)
+	if err != nil {
+		return false, err
+	}
+
+	return binary.BigEndian.Uint16(b) > 0, nil
+}
+
+// Enable implements the api.Charger interface
+func (wb *Dadapower) Enable(enable bool) error {
+	var u uint16
+	if enable {
+		u = 1
+	}
+
+	_, err := wb.conn.WriteSingleRegister(dadapowerRegChargingAllowed+wb.regOffset, u)
+
+	return err
+}
+
+// MaxCurrent implements the api.Charger interface
+func (wb *Dadapower) MaxCurrent(current int64) error {
```

Comment:
Nur ganze Ampere? Sonst wäre MaxCurrentMillis noch schön.

Reply:
ja, geht auch - müsste ich jedoch in der Wallbox Firmware umstellen.

Reply:
Eines der wichtigsten Features.

Reply:
Ok, ist erledigt.